File: mmd-dashboard/lib/validator.ts
// Structural validation engine and export-readiness gate (Sections 27-28, 43).
// Port of mmd_generator/validator.py.

import type { LetterType, ReadinessGate, ValidationIssue } from "./models";
import type { TagManager } from "./tagManager";

const NAME_REF_LEFTOVERS: Array<{ re: RegExp; label: string }> = [
  { re: /\(\s*Title\s*\)/, label: "(Title)" },
  { re: /\(\s*Forename\s*\)/, label: "(Forename)" },
  { re: /\(\s*Surname\s*\)/, label: "(Surname)" },
];

export function findUnmatchedBrackets(text: string): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  const stack: number[] = [];
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (ch === "[") {
      stack.push(i);
    } else if (ch === "]") {
      if (stack.length) {
        stack.pop();
      } else {
        issues.push({
          severity: "ERROR",
          code: "UNMATCHED_CLOSE_BRACKET",
          message: `Unmatched ']' at character offset ${i}.`,
        });
      }
    }
  }
  for (const i of stack) {
    issues.push({
      severity: "ERROR",
      code: "UNMATCHED_OPEN_BRACKET",
      message: `Unmatched '[' at character offset ${i}.`,
    });
  }
  return issues;
}

export function findLeftoverNamePlaceholders(body: string): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  for (const { re, label } of NAME_REF_LEFTOVERS) {
    if (re.test(body)) {
      issues.push({
        severity: "WARNING",
        code: "UNRESOLVED_NAME_PLACEHOLDER",
        message: `Unresolved raw placeholder '${label}' remains in the generated MMD.`,
      });
    }
  }
  return issues;
}

export function checkCrossContextTags(body: string, letterType: LetterType): ValidationIssue[] {
  const otherPrefix = letterType === "MEMBER" ? "dependant" : "member";
  const otherFields = [
    `${otherPrefix}_title_guess`,
    `${otherPrefix}_forename`,
    `${otherPrefix}_surname`,
    `${otherPrefix}_address`,
    `${otherPrefix}_postcode_or_foreign_country`,
  ];
  const issues: ValidationIssue[] = [];
  for (const tag of otherFields) {
    const re = new RegExp(`<%\\s*${tag}\\s*%>`);
    if (re.test(body)) {
      issues.push({
        severity: "ERROR",
        code: "CROSS_CONTEXT_TAG",
        message: `Found '<% ${tag} %>' in a ${letterType} letter -- recipient context mismatch.`,
      });
    }
  }
  return issues;
}

export function validateTagsAndConditionals(tagManager: TagManager): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  for (const tag of tagManager.tags()) {
    if (tag.name && tag.name.includes(" ")) {
      issues.push({
        severity: "ERROR",
        code: "SPACE_IN_TAG_NAME",
        message: `Generated tag '${tag.name}' contains a space -- invalid MMD tag name.`,
      });
    }
    if (tag.status === "Needs Creation") {
      issues.push({
        severity: "WARNING",
        code: "TAG_NEEDS_CREATION",
        message: `Tag '${tag.name}' needs creation (${tag.occurrences} occurrence(s)).`,
      });
    }
    if (tag.status === "Duplicate") {
      issues.push({
        severity: "WARNING",
        code: "DUPLICATE_TAG",
        message: `Tag '${tag.name}' may duplicate another field. ${tag.notes}`,
      });
    }
    if (tag.status === "Review") {
      issues.push({
        severity: "WARNING",
        code: "TAG_REVIEW_REQUIRED",
        message: `Tag '${tag.name || "(unresolved)"}' requires manual review.`,
      });
    }
  }
  for (const cond of tagManager.conditionals()) {
    if (!cond.confirmed) {
      issues.push({
        severity: "WARNING",
        code: "CONDITIONAL_SCOPE_UNCONFIRMED",
        message: `Conditional '${cond.name}' requires scope confirmation before export.`,
      });
    }
  }
  return issues;
}

export function computeGate(issues: ValidationIssue[]): ReadinessGate {
  if (issues.some((i) => i.severity === "ERROR")) return "NOT READY";
  if (issues.some((i) => i.severity === "WARNING")) return "REVIEW REQUIRED";
  return "READY";
}

File: mmd_generator/validator.py
"""Structural validation engine and export-readiness gate (Sections 27-28, 43)."""
from __future__ import annotations

import re
from typing import List

from .models import (
    LetterType,
    ReadinessGate,
    Severity,
    TagStatus,
    ValidationIssue,
)
from .tag_manager import TagManager

_NAME_REF_LEFTOVERS = [
    (re.compile(r"\(\s*Title\s*\)"), "(Title)"),
    (re.compile(r"\(\s*Forename\s*\)"), "(Forename)"),
    (re.compile(r"\(\s*Surname\s*\)"), "(Surname)"),
]


def find_unmatched_brackets(text: str) -> List[ValidationIssue]:
    """Raw character-level ``[``/``]`` balance check -- a low-level safety
    net independent of the token-based conditional parser, catching
    malformed brackets that don't even form a valid ``[...]`` token."""
    issues: List[ValidationIssue] = []
    stack: List[int] = []
    for i, ch in enumerate(text):
        if ch == "[":
            stack.append(i)
        elif ch == "]":
            if stack:
                stack.pop()
            else:
                issues.append(
                    ValidationIssue(
                        severity=Severity.ERROR,
                        code="UNMATCHED_CLOSE_BRACKET",
                        message=f"Unmatched ']' at character offset {i}.",
                    )
                )
    for i in stack:
        issues.append(
            ValidationIssue(
                severity=Severity.ERROR,
                code="UNMATCHED_OPEN_BRACKET",
                message=f"Unmatched '[' at character offset {i}.",
            )
        )
    return issues


def find_leftover_name_placeholders(body: str) -> List[ValidationIssue]:
    issues = []
    for pattern, label in _NAME_REF_LEFTOVERS:
        if pattern.search(body):
            issues.append(
                ValidationIssue(
                    severity=Severity.WARNING,
                    code="UNRESOLVED_NAME_PLACEHOLDER",
                    message=f"Unresolved raw placeholder '{label}' remains in the generated MMD.",
                )
            )
    return issues


def check_cross_context_tags(body: str, letter_type: LetterType) -> List[ValidationIssue]:
    """Section 27: 'incorrect member/dependant references' -- flag protected
    recipient tags belonging to the *other* recipient type appearing in the
    generated MMD (member_id is shared and intentionally excluded)."""
    other_prefix = "dependant" if letter_type == LetterType.MEMBER else "member"
    other_fields = [
        f"{other_prefix}_title_guess",
        f"{other_prefix}_forename",
        f"{other_prefix}_surname",
        f"{other_prefix}_address",
        f"{other_prefix}_postcode_or_foreign_country",
    ]
    issues = []
    for tag in other_fields:
        if re.search(rf"<%\s*{re.escape(tag)}\s*%>", body):
            issues.append(
                ValidationIssue(
                    severity=Severity.ERROR,
                    code="CROSS_CONTEXT_TAG",
                    message=(
                        f"Found '<% {tag} %>' in a {letter_type.value} letter -- "
                        "recipient context mismatch."
                    ),
                )
            )
    return issues


def validate_tags_and_conditionals(tag_manager: TagManager) -> List[ValidationIssue]:
    issues: List[ValidationIssue] = []
    for tag in tag_manager.tags():
        if tag.name and " " in tag.name:
            issues.append(
                ValidationIssue(
                    severity=Severity.ERROR,
                    code="SPACE_IN_TAG_NAME",
                    message=f"Generated tag '{tag.name}' contains a space -- invalid MMD tag name.",
                )
            )
        if tag.status == TagStatus.NEEDS_CREATION:
            issues.append(
                ValidationIssue(
                    severity=Severity.WARNING,
                    code="TAG_NEEDS_CREATION",
                    message=f"Tag '{tag.name}' needs creation ({tag.occurrences} occurrence(s)).",
                )
            )
        if tag.status == TagStatus.DUPLICATE:
            issues.append(
                ValidationIssue(
                    severity=Severity.WARNING,
                    code="DUPLICATE_TAG",
                    message=f"Tag '{tag.name}' may duplicate another field. {tag.notes}",
                )
            )
        if tag.status == TagStatus.REVIEW:
            issues.append(
                ValidationIssue(
                    severity=Severity.WARNING,
                    code="TAG_REVIEW_REQUIRED",
                    message=f"Tag '{tag.name or '(unresolved)'}' requires manual review.",
                )
            )
    for cond in tag_manager.conditionals():
        if not cond.confirmed:
            issues.append(
                ValidationIssue(
                    severity=Severity.WARNING,
                    code="CONDITIONAL_SCOPE_UNCONFIRMED",
                    message=f"Conditional '{cond.name}' requires scope confirmation before export.",
                )
            )
    return issues


def compute_gate(issues: List[ValidationIssue]) -> ReadinessGate:
    if any(i.severity == Severity.ERROR for i in issues):
        return ReadinessGate.NOT_READY
    if any(i.severity == Severity.WARNING for i in issues):
        return ReadinessGate.REVIEW_REQUIRED
    return ReadinessGate.READY
